### PR TITLE
Add new source and push for Bitbucket server using bearer auth 

### PR DESCRIPTION
**Pull request by noorul**
Bitbucket server administrator can disable Basic HTTP Auth scheme. The
other option is to use Bearer scheme.

### CODE REVIEW

**Comment by noorul** [comment]
@asottile Can you please take a look? Looks like test failures are not related.

**Comment by asottile** [comment]
there's a different failure than those that is related to your change -- but hidden by those failures

**Comment by noorul** [comment]
@asottile How do we take it forward?

**Comment by asottile** [comment]
perhaps fix the tests?

**Comment by noorul** [comment]
@asottile Can you take a look now?

File: all_repos/push/bitbucket_server_pull_request.py [hunk 1]
```diff
@@ -29,9 +30,14 @@ def make_pull_request(
         settings: Settings,
         branch_name: str,
 ) -> bitbucket_server_api.Response:
+    if settings.bearer_auth:
+        auth_header = {'Authorization': f'Bearer {settings.app_password}'}
+    else:
+        auth_header = {'Authorization': f'Basic {settings.auth}'}
```
Comment (asottile) [inline all_repos/push/bitbucket_server_pull_request.py:36]: this to me feels like it should be a separate `source`.  `app_password` isn't actually the app password but a bearer token

File: tests/grep_test.py [hunk 2]
```diff
@@ -110,7 +110,7 @@ def _test_grep_color(file_config_files, capsys, *, args=()):
     expected = (
         '\033[1;34m{}\033[m'
         '\033[36m:\033[m'
-        'f'
+        '\033[35mf\033[m'
```
Comment (asottile) [inline tests/grep_test.py:113]: can you split this out to a separate PR -- also this should accept both formats (it's unfortunately sensitive to the git version -- the version on my host doesn't highlight this purple)

File: tests/source/bitbucket_server_test.py [hunk 3]
```diff
@@ -34,10 +34,15 @@ def repos_project_response(mock_urlopen):
...
-        'cool_user', 'app_password', 'bitbucket.domain.com',
-    )
+@pytest.mark.parametrize(
+    ('setting'),
+    (
+        ('cool_user', 'app_password', 'bitbucket.domain.com'),
+        ('cool_user', 'app_password', 'bitbucket.domain.com', None, True),
+    ),
+)
+def test_list_repos(setting):
```
Comment (asottile) [inline tests/source/bitbucket_server_test.py:44]: just make the `Settings` object in the parametrize itself rather than having indirection

also this doesn't demonstrate that your change works at all, it just lazily covers the code

File: all_repos/push/bitbucket_server_pull_request.py (hunk 1 above)
Comment (noorul) [reply all_repos/push/bitbucket_server_pull_request.py:36]: As per the documentation https://confluence.atlassian.com/bitbucketserver/http-access-tokens-939515499.html this password can be used for both `Basic` and `Bearer` auth. I am not sure why we need two separate source.

File: tests/grep_test.py (hunk 2 above)
Comment (noorul) [reply tests/grep_test.py:113]: Done via https://github.com/asottile/all-repos/pull/219

File: tests/source/bitbucket_server_test.py (hunk 3 above)
Comment (noorul) [reply tests/source/bitbucket_server_test.py]: Used Settings object directly. I was looking for tests for request validation the repo and I could not find any. Am I missing something?

**Comment by noorul** [comment]
@asottile Incorporated review comments, please take a look.

File: all_repos/push/bitbucket_server_pull_request.py (hunk 1 above)
Comment (asottile) [reply all_repos/push/bitbucket_server_pull_request.py:36]: they are fundamentally different auth models -- a bearer token does not have a username.  they should be two separate sources rather than a boolean trap

Comment (noorul) [reply all_repos/push/bitbucket_server_pull_request.py]: @asottile Alright, I created separate source and push for this.

**Comment by noorul** [comment]
@asottile Can you take a look?

File: README.md [hunk 4]
```diff
@@ -456,6 +471,17 @@ Pushes the branch to `origin` and then creates a Bitbucket pull request for the
...
+### `all_repos.push.bitbucket_server_pull_request_bearer_auth`
+
+Pushes the branch to `origin` and then creates a Bitbucket pull request for the branch using bearer auth scheme.
+
+#### Required `push_settings`
+
+- `base_url`: the Bitbucket server URL (eg `bitbucket.domain.com`)
+- `token`: the personal access token which the user will log in
+    - Create a personal access token within your [account settings](https://bitbucket.domain.com/plugins/servlet/access-tokens/manage).
+    - We need the scope: Repositories -> Read
```
Comment (asottile) [inline README.md:483]: presumably you'd need higher than `read` to push ?

File: all_repos/push/bitbucket_server_pull_request_bearer_auth.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
+from __future__ import annotations
+
+from typing import NamedTuple
+
+from all_repos.push.bitbucket_server_pull_request import _push
```
Comment (asottile) [inline all_repos/push/bitbucket_server_pull_request_bearer_auth.py:5]: you shouldn't import underscored names.

also this is over-DRY, don't be afraid of copying a little bit of code vs. making spaghetti

File: README.md (hunk 4 above)
Comment (noorul) [reply README.md:483]: No, Read scope is good enough as I have been using it for a long time.

File: all_repos/push/bitbucket_server_pull_request_bearer_auth.py (hunk 5 above)
Comment (noorul) [reply all_repos/push/bitbucket_server_pull_request_bearer_auth.py:5]: I will refactor common code into a separate package. I don't think it is too much DRY. In fact, the only difference is auth scheme.

Comment (noorul) [reply all_repos/push/bitbucket_server_pull_request_bearer_auth.py]: @asottile Can you please take a look now?

**Comment by noorul** [comment]
@asottile I incorporated review comments, can you take a look.

**Comment by noorul** [comment]
Can you help me to understand what further changes will make this PR ready for merge?

**Comment by asottile** [comment]
it's bad form to ping incessantly

I am a busy person

chill out and I'll get to it, keep pinging and I'll be less and less incentived to look at it

File: README.md (hunk 4 above)
Comment (asottile) [reply README.md:483]: how do you make a push with a readonly token?

Comment (noorul) [reply README.md]: Updated the patch.
